components/raftstore/src/store/peer.rs

```diff
@@ -3764,20 +3829,54 @@ where
                 normal_peers.insert(peer_id);
             }
             if let Some(pr) = self.raft_group.raft.prs().get(peer_id) {
-                next_idxs.push((peer_id, pr.next_idx, usage));
+                // status 3-normal, 2-almostfull, 1-alreadyfull, only for simplying the sort func belowing.
+                let mut status = 3;
+                if let Some(usg) = usage {
+                    status = match usg {
+                        DiskUsage::Normal => 3,
+                        DiskUsage::AlmostFull => 2,
+                        DiskUsage::AlreadyFull => 1,
+                    };
+                }
+
+                if !self.down_peer_ids.contains(&peer_id) {
+                    next_idxs.push((peer_id, pr.next_idx, usage, status));
+                    if min_peer_index > pr.next_idx {
+                        min_peer_index = pr.next_idx;
+                    }
+                }
             }
         }
 
+        if self.has_region_merge_proposal && min_peer_index >= self.region_merge_proposal_index {
+            self.has_region_merge_proposal = false;
+        }
+
         if normal_peers.len() == peers_len {
             return;
         }
 
+        self.latest_majority_set_update_on_disk_full = cur;
+        // Reverse sort peers based on `next_idx`, `usage` and `store healthy status`, then try to get a potential quorum.
+        next_idxs.sort_by(|x, y| {
+            if x.3 == y.3 {
+                y.1.cmp(&x.1)
+            } else {
+                y.3.cmp(&x.3)
+            }
+        });
+
         let raft = &mut self.raft_group.raft;
         self.disk_full_peers.majority = !raft.prs().has_quorum(&normal_peers);
-        for &(peer, _, usage) in &next_idxs {
+        for &(peer, _, usage, ..) in &next_idxs {
             if let Some(usage) = usage {
-                self.disk_full_peers.peers.insert(peer, (*usage, false));
-                raft.adjust_max_inflight_msgs(peer, 0);
+                if self.has_region_merge_proposal && !matches!(*usage, DiskUsage::AlreadyFull) {
+                    self.disk_full_peers.peers.insert(peer, (*usage, true));
+                    raft.adjust_max_inflight_msgs(peer, 1);
```

Comment:
Why set to 1? if we want to replica the merge proposal to the peer as soon as possible, we should set it to a higher value.

Reply:
Set to 1 is only to simply the judge logic.
Because the logic here is under the heartbeat context, we don't know what type of the future cmd is. 
The appropriate processing time should be when the leader processes the proposal
However, it is currently recommended to keep the current implementation.

Reply:
`inflight_msgs` is not just for future cmd but also lagging logs.

Reply:
OK, I agree, set all to the `cfg.raft_max_inflight_msgs`